src/link/script.y

```diff
@@ -0,0 +1,479 @@
+		// Basically yywrap().
+		lexerStack.pop_back();
+		if (!lexerStack.empty()) {
+			goto try_again;
+		}
+		return yy::parser::make_YYEOF();
+	} else if (isNewline(c)) {
+		// Handle CRLF.
+		if (c == '\r' && context.file.sgetc() == '\n') {
+			context.file.sbumpc();
+		}
+		return yy::parser::make_newline();
+	} else if (c == '"') {
+		std::string str;
+
+		for (c = context.file.sgetc(); c != '"'; c = context.file.sgetc()) {
+			if (c == LexerStackEntry::eof || isNewline(c)) {
+				scriptErrorSimple(context, "Unterminated string");
+				break;
+			}
+			context.file.sbumpc();
+			if (c == '\\') {
+				c = context.file.sgetc();
+				if (c == LexerStackEntry::eof || isNewline(c)) {
+					scriptErrorSimple(context, "Unterminated string");
+					break;
+				} else if (c == 'n') {
+					c = '\n';
+				} else if (c == 'r') {
+					c = '\r';
+				} else if (c == 't') {
+					c = '\t';
+				} else if (c != '\\' && c != '"') {
+					scriptError(context, "Cannot escape character %s", printChar(c));
+				}
+				context.file.sbumpc();
+			}
+			str.push_back(c);
+		}
+		context.file.sbumpc(); // Consume the closing quote.
+
+		return yy::parser::make_string(std::move(str));
+	} else if (c == '$') {
+		c = context.file.sgetc();
+		if (!isHexDigit(c)) {
+			scriptErrorSimple(context, "No hexadecimal digits found after '$'");
+			return yy::parser::make_number(0);
+		}
+
+		uint32_t number = parseHexDigit(c);
+		context.file.sbumpc();
+		for (c = context.file.sgetc(); isHexDigit(c); c = context.file.sgetc()) {
+			number = number * 16 + parseHexDigit(c);
+			context.file.sbumpc();
+		}
+		return yy::parser::make_number(number);
+	} else if (c == '%') {
+		c = context.file.sgetc();
+		if (!(c >= '0' && c <= '1')) {
+			scriptErrorSimple(context, "No binary digits found after '%%'");
+			return yy::parser::make_number(0);
+		}
+
+		uint32_t number = c - '0';
+		context.file.sbumpc();
+		for (c = context.file.sgetc(); c >= '0' && c <= '1'; c = context.file.sgetc()) {
+			number = number * 2 + (c - '0');
+			context.file.sbumpc();
+		}
+		return yy::parser::make_number(number);
+	} else if (c >= '0' && c <= '9') {
+		uint32_t number = c - '0';
+		for (c = context.file.sgetc(); c >= '0' && c <= '9'; c = context.file.sgetc()) {
+			number = number * 10 + (c - '0');
+		}
+		return yy::parser::make_number(number);
+	} else if (isIdentChar(c)) { // Note that we match these *after* digit characters!
+		std::string ident;
+		auto strCaseCmp = [](char cmp, char ref) {
+			// `locale::classic()` yields the "C" locale.
+			assert(!std::use_facet<std::ctype<char>>(std::locale::classic())
+			       .is(std::ctype_base::lower, ref));
+			return std::use_facet<std::ctype<char>>(std::locale::classic())
+			       .toupper(cmp) == ref;
+		};
+
+		ident.push_back(c);
+		for (c = context.file.sgetc(); isIdentChar(c); c = context.file.snextc()) {
+			ident.push_back(c);
+		}
+
+		for (SectionType type : EnumSeq(SECTTYPE_INVALID)) {
+			if (std::equal(ident.begin(), ident.end(),
```

Comment:
If we can't have `ranges`, how about a `#define RANGE(s) std::begin(s), std::end(s)` macro? `grep -rP '(\S+)\.begin\(\), \1\.end\(\)' src` finds 32 cases we could use it so far.

Reply:
Hmm, sure, but I'd like to have that as a separate PR so we keep the eventual squashed commit clean enough.

Reply:
Files as #1267.